### PR TITLE
feat:  add meta with codes on warnings to allow conditional logging

### PR DESCRIPTION
**Pull request by delta-9**
#### Checklist

- [x] only relevant code is changed (make a diff before you submit the PR)
- [x] run tests `npm run test`
- [ ] tests are included
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/i18next/.github/blob/master/CONTRIBUTING.md)

#### Checklist (for documentation change)

- [ ] only relevant documentation part is changed (make a diff before you submit the PR)
- [ ] motivation/reason is provided
- [ ] commit message and code follows the [Developer's Certification of Origin](https://github.com/i18next/.github/blob/master/CONTRIBUTING.md)

### CODE REVIEW

**Comment by adrai** [comment]
not necessary, since you can now inject your own logger: https://github.com/i18next/react-i18next-gitbook/pull/150#issuecomment-2564710238

**Comment by delta-9** [comment]
@adrai I saw your changes just now, cool!, this PR now just adds meta so I can filter what to log or not

**Comment by adrai** [comment]
I suggest, you move this logic out to your custom logger... else this will bloat the code size even more... //cc @VIKTORVAV99

**Comment by delta-9** [comment]
> I suggest, you move this logic out to your custom logger... else this will bloat the code size even more... //cc @VIKTORVAV99

fair enough, I reworked the changes to includes only what is needed to do that from the logger

**Comment by adrai** [comment]
Do you really need extra error codes?

**Comment by delta-9** [comment]
> Do you really need extra error codes? The whole i18next core code base has no error codes...

the error codes is to be able to identify the warnings in the code.
it also adds the i18nkey to the warning logged so its easier to find the culprit trans.

the i18nkey also allow to add a logic to warnOnce per key in the custom logger, now it warn on each renders.

for the code I could do some string regexp on the messages to figure out the cause of the error,
and for the i18nkey, Its also possible to debug following the stack trace and put some trace until I find the right call.

**Comment by VIKTORVAV99** [comment]
Maybe there are benefits from this but I would like you to consider two things:

First this will increase the overhead and bundle size for all users of i18next, I've been trying to push this bundle down as it's one of the bigger parts of our app we can't change at the moment.

Second, I would almost call this a breaking change, partly as it changes the log outputs which means all tests or error identifying software such as sentry will either fail or log these as new errors. Secondly it feels like this is just duplicating the error message. If we do this I would switch to something what react does and only log the error codes with a link to look up what they mean. But that requires other changes as well.

That is just my thought on the matter though but I would really like to avoid duplicating any information or functionality for the sake of making it a little more convenient for one project or developer.

**Comment by rawpixel-vincent** [comment]
I replaced the error map with a tuple, and reduced duplicates, it's adding 201 bytes to `react-i18next.min.js` build compared to current master

File: src/utils.js [hunk 1]
```diff
@@ -1,20 +1,33 @@
-export const warn = (i18n, ...args) => {
+/** @template {string} T @param {...T} s @return {T[]} */
+const tup = (...s) => s;
```
Comment (adrai) [inline src/utils.js:2]: The whole ERR_CODES and tup code can be removed now, right?

**Comment by delta-9** [comment]
If you ask me, I think it's needed for sanity.
But the code can run without it and the type accept any string as code.

**Comment by delta-9** [comment]
good news @adrai I moved the error code to a type declaration file so its not in the code.

the minify js is now 20 bytes less than master 🎉

**Comment by delta-9** [comment]
added types for along an example of usage in the logger plugin

**Comment by adrai** [comment]
TS tests seems to fail: https://github.com/i18next/react-i18next/actions/runs/12542746999/job/34976361013?pr=1826

**Comment by delta-9** [comment]
it should be fixed, but I can't make the test fail locally, npm run test pass

**Comment by adrai** [comment]
it's included in v15.4.0

**Pull request merged**